project/chu-siang/README.rst

```diff
@@ -0,0 +1,241 @@
+               dt = time.time() - self.t0
+               if device == "cuda":
+                   peak = torch.cuda.max_memory_allocated() / (1024 ** 3)
+                   print(f"[{self.name}] {dt:.2f}s, peak={peak:.2f}GB")
+               else:
+                   print(f"[{self.name}] {dt:.2f}s")
+
+       model.eval()
+       with torch.no_grad():
+           autocast = (
+               torch.cuda.amp.autocast(dtype=torch.float16)
+               if device == "cuda" else nullcontext()
+           )
+           with Timed("train(grpo)"), autocast:
+               trainer.train()
+
+   if __name__ == "__main__":
+       main()
+
+C++: Fused sampling (temperature + softmax + multinomial) skeleton
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. code-block:: cpp
+
+   // file: fused_sample.cpp
+   #include <torch/extension.h>
+
+   torch::Tensor fused_sample(torch::Tensor logits, double temperature) {
+       TORCH_CHECK(logits.is_cuda(), "logits must be CUDA");
+       auto scaled = logits / temperature;
+       auto probs  = torch::softmax(scaled, -1);
+       auto ids = torch::multinomial(probs, 1);
+       return ids;
+   }
+
+   PYBIND11_MODULE(TORCH_EXTENSION_NAME, m) {
+       m.def("fused_sample", &fused_sample,
+             "Fused temperature-softmax-sample");
+   }
+
+Build script (setup.py)
+~~~~~~~~~~~~~~~~~~~~~~~
+
+.. code-block:: python
+
+   from setuptools import setup
+   from torch.utils.cpp_extension import CppExtension, BuildExtension
+
+   setup(
+       name="fused_sample_ext",
+       ext_modules=[
+           CppExtension("fused_sample_ext", ["fused_sample.cpp"]),
+       ],
+       cmdclass={"build_ext": BuildExtension},
+   )
+
+Engineering Infrastructure
+--------------------------
+
+- **Automatic build system**: Python env (``requirements.txt``), Dockerfile
+  (CUDA 12.x + PyTorch + TRL), C++ extension auto-build.
+- **Version control**: ``main`` for stable, ``feat/*`` for experiments, PRs
+  gated by lint/tests.
+- **Testing**: CPU smoke tests, GPU V100 short GRPO runs, tokens/sec + VRAM
+  benchmarks.
+- **Documentation**: ``README.md``, ``docs/`` (Sphinx), inline docstrings.
+- **CI**: GitHub Actions (CPU), optional self-hosted GPU runner.
+- **Optimization playbook**:
+
+  - Python: ``model.eval()``, ``no_grad``, FP16 autocast, length bucketing,
+    vectorized reward.
+  - C++: fused sampling kernel, CUDA Graphs, KV pooling.
+  - Numerical: log-sum-exp for KL, clamped probs.
+  - System: stage profiling, Accelerate/DeepSpeed A/B tests.
+
+Schedule
+--------
+
+- **Planning phase (8 weeks from 09/22 to 11/09)**
+- Week 1 (09/22): Repo genearte, baseline GRPO run on V100.
+- Week 2 (09/30): Profiling hooks,Docker setup, tune gen_kwargs, enforce FP16 inference.
+- Week 3 (10/08): Implement C++ ``fused_sample_ext``, A/B test vs Python.
+- Week 4 (10/15): Integrate Reward Model via TRL RewardTrainer, vectorize
+  reward.
+- Week 5 (10/22): Multi-GPU scaling with Accelerate/DeepSpeed.
```

Comment:
Please specify how do you write GPU/C++ code.

Failing to write C++ code yourself will result in poor grade.